[2단계 - 장바구니 기능] 제이미(임정수) 미션 제출합니다.

JJ503 opened the pull request:
안녕하세요 춘식!
1단계 미션에 이어 2단계 미션도 PR 리뷰 요청합니다.
고민과 잘 안 되는 부분들이 많아 생각보다 리뷰 요청이 늦어졌습니다 😥

의문이 든 사항에 대해선 코드에 코멘트를 작성해 두었습니다.
의문 사항 외에도 여러 피드백과 고민해 보면 좋을 것들에 대해 마구 던져주시면 감사하겠습니다!

또한, 1단계에서 주신 피드백에 대한 답변 및 추가 의문 사항에 대해 1단계 PR 코멘트로 달아두었습니다.
이에 대해서도 확인해 주시면 감사하겠습니다! ([1단계  PR 바로가기](https://github.com/woowacourse/jwp-shopping-cart/pull/237))

이번 리뷰도 잘 부탁드립니다!
감사합니다.

JJ503 committed:
refactor: autowired 어노테이션 생략가능한 부분 생략

JJ503 committed:
refactor: Controller 어노테이션을 RestController로 수정

JJ503 committed:
refactor: 예외 advice 중복 코드 제거

JJ503 committed:
refactor: Validated대신 Valid로 변경

JJ503 committed:
refactor: 예외처리 추가

JJ503 committed:
refactor: 테스트 코드 적절하게 수정

JJ503 committed:
docs: 2단계 기능 요구사항 목록 추가

JJ503 committed:
feat: UserDao 저장 기능 추가

JJ503 committed:
feat: UserService 저장 기능 추가

JJ503 committed:
feat: UserController에 POST /user 요청 매핑

JJ503 committed:
feat: UserDao 전체 및 특정 user 조회 기능 추가

JJ503 committed:
feat: UserService 전체 및 특정 user 조회 기능 추가

JJ503 committed:
feat: UserController에 GET /user/{userId} 요청 매핑

JJ503 committed:
refactor: User 빌더 패턴 적용

JJ503 committed:
refactor: User 찾은 값이 null도 가능하도록 수정

JJ503 committed:
refactor: UserDao 수정 기능 추가

JJ503 committed:
refactor: UserService 수정 기능 추가

JJ503 committed:
refactor: UserController에 PUT /user 요청 매핑

JJ503 committed:
docs: 사용자 기능 구현 완료 체크

JJ503 committed:
feat: 사용자 설정 페이지 연동

JJ503 committed:
feat: CartDao 저장 기능 추가

JJ503 committed:
feat: CartService 저장 기능 추가

JJ503 committed:
feat: CartController POST /carts 요청 매핑

JJ503 committed:
refactor: 클래스명 수정

ItemViewController -> ViewController

JJ503 committed:
refactor: CartRequest 필드 수정

userId 제거

JJ503 committed:
refactor: user의 email을 unique 속성으로 변경

JJ503 committed:
feat: UserDao 이메일로 사용자 찾기 기능 추가

JJ503 committed:
feat: CartDao 사용자 아이디에 따른 조회 기능 추가

JJ503 committed:
feat: CartService 사용자 아이디에 따른 조회 기능 추가

- CartResponse 반환 dto 생성
- AuthInfo 계정 요청 dto 생성
- Service에서 이메일이 아닌 AuthInfo를 받도록 수정

JJ503 committed:
feat: /carts/** 요청 시 로그인 확인 기능 추가

JJ503 committed:
feat: CartController GET /carts 요청 매핑

- 장바구니 CRUD API에 맞게 html, js 코드 변경

JJ503 committed:
feat: 장바구니 페이지 연동 기능 추가

JJ503 committed:
refactor: mapping 어노테이션 '/' 제거

JJ503 committed:
docs: 장바구니 수정 기능 목록 제거

장바구니의 경우 수정할 부분이 없을 것이라 판단해 제거

JJ503 committed:
refactor: auth, config, controller 내부 패키지 수정

JJ503 committed:
refactor: CartController 적절한 메서드명으로 수정

JJ503 committed:
feat: CartDao 삭제 기능 추가

JJ503 committed:
feat: CartService 삭제 기능 추가

JJ503 committed:
feat: CartController DELETE /carts/{id} 요청 매핑

- carts.js 삭제 요청 장바구니 CRUD API에 맞게 변경

JJ503 committed:
fix: 요구사항에 맞도록 장바구니 추가 기능 수정

POST /carts 요청 시 product가 아닌 productId가 전달되도록 수정

JJ503 committed:
refactor: 도메인의 각 필드 랩핑 및 검증 로직 추가

JJ503 committed:
fix: 테스트 수행 시 오류 발생 해결

- 외래키 참조로 삭제 시 오류가 발생하는 문제 해결
- WebMvcTest 수행 시 특정 컨트롤러에 대해서만 수행되도록 수정

JJ503 committed:
refactor: Service 클래스에 @Transactional 설정

데이터 무결성 및 일관성 보장을 위해

JJ503 committed:
refactor: NamedParameterJdbcTemplate를 사용하도록 수정

JJ503 committed:
feat: 존재하는 장바구니 아이디인지 확인하는 기능 추가

JJ503 committed:
feat: 각 도메인 동등성 재정의

JJ503 committed:
feat: UserDao 삭제 기능 추가

JJ503 committed:
test: Userdao 단위 테스트 추가

JJ503 committed:
refactor: ItemDaoTest displayName 맞춤법 수정

JJ503 committed:
feat: UserService 사용자 제거 기능 추가

JJ503 committed:
feat: 반환 dto 동등성 재정의

JJ503 committed:
refactor: todo 제거

JJ503 committed:
fix: 중복되는 이메일 저장 시 예외 처리

JJ503 committed:
test: UserService 단위 테스트 추가

JJ503 committed:
refactor: 요청 url을 user에서 users로 수정

JJ503 committed:
refactor: 요청 url을 user에서 users로 수정

JJ503 committed:
refactor: ItemControllerTest displayname 명확하게 수정

JJ503 committed:
feat: DELETE /users/{id} 요청 매핑

JJ503 committed:
refactor: 반환 location 수정

JJ503 committed:
test: UserController 단위 테스트 추가

JJ503 committed:
refactor: ItemControllerTest 맞춤법 수정

JJ503 committed:
refactor: CartDao의 findBy() 반환 타입 수정

JJ503 committed:
feat: CartDao 단위 테스트 추가

JJ503 committed:
feat: CartServiceTest 단위 테스트 추가

JJ503 committed:
feat: CartController 단위 테스트 추가

JJ503 committed:
refactor: UserRequest 예외 메시지 추가

JJ503 committed:
test: ViewController 및 UserController 통합 테스트 추가

JJ503 committed:
feat: 아이템 존재 여부 확인 기능 추가

JJ503 committed:
test: CartController 통합 테스트 추가

JJ503 committed:
test: Email 단위 테스트 추가

JJ503 committed:
test: Password 단위 테스트 추가

JJ503 committed:
feat: User 검증 로직 추가

JJ503 committed:
test: User 단위 테스트 추가

JJ503 commented on src/main/java/cart/config/WebMvcConfiguration.java:
```diff
@@ -0,0 +1,31 @@
+public class WebMvcConfiguration implements WebMvcConfigurer {
+    @Override
+    public void addViewControllers(ViewControllerRegistry registry) {
+        registry.addViewController("/cart")
+                .setViewName("cart");
```
`/cart`로 requestMapping 후 `cart.html`을 반환하는 과정에서 별다른 로직이 필요 없기에 Configuration 클래스 내에서 위와 같은 방법을 통해 url과 view를 매핑해 주었습니다.
실제로도 이렇게 많이 사용되나요?
혹은 추후 로직이 생길 수도 있기 때문에 확장성 및 통일성 등을 생각해 `ViewController`에서 함께 관리해 주는 것이 좋을까요?

bimppap replied:
실무에서는 프론트에서 view 관련된 부분을 전담하고 있기 때문에 백엔드에서 view를 다룰 일은 거의 없습니다. 하지만 다룰 일이 생긴다면 한쪽에서 전부 관리하도록 하는게 관리하기엔 편합니다!

JJ503 committed:
refactor: mapping url의 / 제거

JJ503 commented on src/main/java/cart/controller/CartController.java:
```diff
@@ -0,0 +1,49 @@
+        this.cartService = cartService;
+    }
+
+    @PostMapping("{itemId}")
+    public ResponseEntity<Void> addCart(@BasicAuthenticationPrincipal AuthInfo authInfo,
+                                        @PathVariable final Long itemId) {
```
현재 `addCart()`의 PathVariable의 경우 상품의 아이디를 받습니다.
반면, `deleteCart()`의 PathVariable의 경우 장바구니의 아이디를 받습니다.
다른 정보를 받음에도 url 요청만 보았을 때는 서로 다른 정보를 받는지 알기 어려운 상태입니다.
이런 경우가 크게 상관없는지 혹은 다른 개발자들에게 혼돈을 주는지 궁금합니다!

bimppap replied:
프론트엔드와 명세 요구사항 공유가 잘 되었다면 이 부분은 걱정하지 않아도 됩니다. 👍 (공유가 안 되었다면...🙄) 백엔드 팀 내에서는 헷갈리지 않도록 지금 PathVariable에 적용한 것처럼 id 대신 itemId, orderId 등 어디의 id의 명시하는게 좋습니다.

JJ503 committed:
feat: UserRequest 검증 추가

JJ503 commented on src/main/java/cart/dao/CartDao.java:
```diff
@@ -0,0 +1,68 @@
+        return (Long) keyHolder.getKey();
+    }
+
+    public List<CartData> findAll(final Long userId) {
+        final String sql = "SELECT carts.id, items.name, items.image_url, items.price " +
+                "FROM items JOIN carts " +
+                "ON items.id = carts.item_id " +
+                "WHERE carts.user_id = :user_id";
+        MapSqlParameterSource param = new MapSqlParameterSource("user_id", userId);
+        return namedParameterJdbcTemplate.query(sql, param, cartDataRowMapper);
+    }
+
```
CartDao의 `findAll()`를 확인해 보면 현재 서비스에서 필요한 정보(장바구니의 id, 상품의 이름, 이미지 url, 가격)만 가져오는 것을 확인할 수 있습니다.
그런데 추후 요구사항이 변경되어 다른 정보들을 가져오는 경우가 발생할 수도 있습니다.
이런 경우를 대비하여 carts, items(혹은 users도)의 테이블을 조인한 모든 정보를 가져오는 것이 좋을지 궁금합니다!

bimppap replied:
요구사항을 미리 대비하여 쿼리를 크게 짜놓을 필요는 없습니다. 오히려 성능이 느려질 수도 있고요. 객체의 경우는 유지보수가 어렵기 때문에 어느 정도의 확장성을 가지고 가는게 좋지만 쿼리는 확장의 굳이 그럴 필요가 없어서 요구사항에만 충실해도 될 것 같네요 🙂

JJ503 replied:
쿼리의 경우는 확장까지 굳이 생각하지 않는군요..!
쿼리의 경우는 비용이 들거나 불필요한 데이터 혹은 객체가 생기기 때문에 그런 것일까요?

JJ503 committed:
feat: Cart 검증 추가

JJ503 committed:
test: Cart 단위 테스트 추가

JJ503 commented on src/main/java/cart/service/CartService.java:
```diff
@@ -0,0 +1,77 @@
+@Transactional
+public class CartService {
+
+    private final CartDao cartDao;
+    private final UserDao userDao;
+    private final ItemDao itemDao;
+
+    public CartService(CartDao cartDao, UserDao userDao, ItemDao itemDao) {
+        this.cartDao = cartDao;
+        this.userDao = userDao;
+        this.itemDao = itemDao;
+    }
```
다른 Service 클래스와 다르게 `CartService` 클래스에서는 세 가지 dao를 모두 사용하고 있습니다.
하지만 UserDaod와 ItemDao는 각각의 서비스에서 사용되고 있는 상태입니다.
그래서 CartService에서도 다시 사용하는 것이 괜찮은지 궁금합니다.
혹은 서비스들을 합치는 것이 더 나을까요?

현재 제가 서비스를 나눈 기준은 테이블 별로 나눈 상태입니다.
춘식의 경우 서비스를 나누는 기준이 어떻게 되는지 궁금합니다!

bimppap replied:
한 서비스에서 쓰이는 dao를 다른 서비스에서 사용하는 건 괜찮습니다. 완전히 같은 기능이라면 서비스에서 다른 서비스를 호출해 사용해도 되지만 이 경우 순환참조가 일어나지 않도록 조심해야 합니다. find~ 처럼 핵심 기능이 아닌 검증의 용도라면 재사용을 하기도 합니다.

서비스는 프로젝트 크기에 따라 나누는 기준을 다르게 잡아도 됩니다. 지금은 도메인/엔티티 단위로 서비스를 작성하지만 서비스가 작다면 하나로 합쳐도 되고, 크다면 더 작게 쪼개도 됩니다.

저희 팀 같은 경우는 작은 도메인은 하나의 서비스를 두고 서비스 내에서 큰 분량을 가진 주제는 다른 서비스로 갈라지기도 합니다. OrderSearchService(조회), OrderService(저장, 업데이트, 삭제) 같은 식으로요. 테이블은 dao에서 조인될 때가 많아서 테이블보다는 도메인으로 나누는 편입니다 🙂

JJ503 replied:
답변 감사합니다!
해당 미션은 작은 서비스라 엔티티 단위로 쪼개도 괜찮지만 다른 미션에서 join을 사용하게 된다면 도메인별로 분리하는 연습도 해봐야겠군요!
cartDao에 join한 데이터를 넘기는 것과 cart 테이블의 한 행 자체를 반환하는 rowMapper가 존재해 서로 다른 도메인을 반환하는데 이 정도는 괜찮을까요? ([cartDao 바로 가기](https://github.com/woowacourse/jwp-shopping-cart/pull/352/files#r1194467418))

JJ503 commented on src/main/java/cart/controller/ItemController.java:
```diff
@@ -41,24 +38,24 @@ public ResponseEntity<List<ItemResponse>> loadAllItem() {
     }
 
-    @PutMapping("/{itemId}")
+    @PutMapping("{itemId}")
```
수정 시 사용할 수 있는 메서드는 Put과 Patch가 있는 것으로 알고 있습니다.
일단 원래 사용한 put을 사용했는데, 춘식의 경우 둘 중 어떤 메서드를 더 선호하시는지 궁금합니다!

bimppap replied:
put과 patch는 용도가 비슷하지만 엄연히 따지자면 구분이 되기 때문에 선호 여부를 떠나 적절한 http 메서드를 적용합니다. 이참에 put 과 patch의 차이를 한번 정리해보는건 어떨까요? 😎

JJ503 replied:
put의 경우는 수정할 데이터가 존재한다면 수정을, 없다면 해당 데이터를 추가하는 방식이군요. 그렇기에 put으로 요청할 경우 수정하고 싶은 데이터만이 아니라 생성 시와 동일한 형식으로 정보를 넘겨주어야 합니다.
반면, patch의 경우 부분적인 수정을 적용하기 위한 메서드로 모든 정보가 아닌 수정하려는 정보만 전달해 주면 됩니다. 그래서 제 경우 존재하는 데이터의 특정 부분만 수정만 하길 원하는 것이므로 patch가 더 잘 어울리겠네요!

그런데 아직 js에서 수정된 값만 보내주는 방법을 잘 모르겠네요… 🥲
또한, patch를 통해 값의 일부만 수정하는 것과 한 행의 모든 데이터를 덮어 씌우는 것이 많이 차이가 나는지도 궁금합니다.
결국 객체를 수정해 이를 dao에 넘겨주는 것으로 보이는데 이는 결국 한 행을 덮어 씌우는 것을 의미합니다.
그렇다면, 일부만 수정하다 null이 실수로 저장되는 것보다 put을 통해 모든 데이터를 가져와 수정하는 방법이 더 안전하지 않을까?라는 생각이 드는데 이는 크게 상관없는지, 그래도 개념적으로 통용해야 하므로 patch가 더 좋은 것인지 궁금합니다!

JJ503 committed:
refactor: 로그 레벨 분리

JJ503 committed:
refactor: 메서드 명 수정

handleWarnException으로 수정

bimppap commented on src/main/java/cart/auth/interceptor/AuthorizationException.java:
```diff
@@ -0,0 +1,14 @@
+    public AuthorizationException() {
+    }
+
+    public AuthorizationException(String message) {
+        super(message);
+    }
```
이건 사용하지 않고 있군요?

bimppap replied:
이 에러가 나면 ControllerAdvice에서 핸들링을 어떻게 할까요?

JJ503 replied:
해당 에러가 발생하게 되면 ControllerAdvice에서 핸들링하지 않고 있기에 500에러가 발생하는군요…!
인증과 관련된 문제는 입력과 동일하게 허용되는 예외 상황이라고 생각해 log.info로 동일하게 처리했습니다.

bimppap commented on src/main/java/cart/auth/interceptor/LoginInterceptor.java:
```diff
@@ -0,0 +1,18 @@
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
+        String accessToken = request.getHeader("Authorization");
+        if (accessToken == null) {
+            throw new AuthorizationException();
```
여기서만 쓰이는 거 같은데 따로 메세지를 안 보내줘도 되는 걸까요? 👀

JJ503 replied:
예외 메시지를 설정하는 것이 더 좋겠네요!
수정했습니다!

bimppap commented on src/main/java/cart/auth/resolver/BasicAuthenticationPrincipalArgumentResolver.java:
```diff
@@ -0,0 +1,26 @@
+
+import javax.servlet.http.HttpServletRequest;
+
+public class BasicAuthenticationPrincipalArgumentResolver implements HandlerMethodArgumentResolver {
```
이건 컴포넌트로 등록하지 않아도 될까요?

JJ503 replied:
제 경우 `BasicAuthenticationPrincipalArgumentResolver`를 매번 객체를 새로 생성해서 사용했는데 빈에 등록하면 그럴 필요 없이 생성된 객체를 가져다 사용할 수 있겠네요!
해당 부분 적용해 보았습니다.

또한, `LoginInterceptor`와 `BasicAuthorizationExtractor`도 동일하게 처리해 주었습니다.
제가 잘 이해한 것인지 확인해 주시면 감사하겠습니다. ([로직 바로 가기](https://github.com/woowacourse/jwp-shopping-cart/pull/352/files#r1195139249))

bimppap commented on src/main/java/cart/controller/dto/auth/AuthInfo.java:
```diff
@@ -0,0 +1,19 @@
+    private String email;
+    private String password;
+
+    public AuthInfo(String email, String password) {
```
AuthInfoDto라고 dto임을 명시해도 되지 않을까요? 👀

JJ503 replied:
Entity 혹은 도메인과 헷갈리지 않기 위해선 명시하는 것이 좋겠네요!

bimppap commented on src/main/java/cart/auth/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,36 @@
+        String header = request.getHeader(AUTHORIZATION);
+
+        if (header == null) {
+            return null;
```
인증이 안 된 경우인데 에러를 던져야 하는 거 아닐까요?

JJ503 replied:
맞습니다..! 예외를 던져주도록 수정했습니다!

bimppap commented on src/main/java/cart/auth/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,36 @@
+            String email = credentials[0];
+            String password = credentials[1];
+
+            return new AuthInfo(email, password);
```
<img width="559" alt="image" src="https://user-images.githubusercontent.com/48675973/236916951-d91d6ff8-8554-4f52-9be8-b92b95fa1f99.png">

이 요청에 대한 응답에 대해 어떻게 생각하시나요?

JJ503 replied:
잘못된 인증 값임에도 존재하지 않는 사용자라는 예외 메시지는 적절하지 않다고 생각합니다.
그래서 base64 인코딩 값으로써 유효한지에 대해 검증하는 로직을 추가해 보았습니다. ([로직 바로 가기](https://github.com/woowacourse/jwp-shopping-cart/pull/352/files#r1195141249))

검증 규칙은 다음과 같습니다.
1. 문자열을 띄어쓰기로 구분했을 때 1개 이상의 값이 존재하는가?
2. 문자열에 알파벳 (A-Z, a-z), 숫자 (0-9), 특수 문자 (+, /, =)의 값만 가지고 있는가?
3. 문자열의 길이가 3에 나누어 떨어지는가?

bimppap commented on src/main/java/cart/controller/ExceptionItemControllerAdvice.java:
```diff
@@ -1,35 +1,24 @@
 
-@ControllerAdvice
+@RestControllerAdvice
 public class ExceptionItemControllerAdvice {
```
클래스명 한번 고민해보면 좋겠습니다!

JJ503 replied:
Controller 외 모든 예외에 대한 처리를 해주는 ExceptionAdvice이기 때문에 GlobalExceptionAdvice로 변경해 보았습니다.
괜찮을까요?

bimppap replied:
기존의 클래스명은 특정 도메인에 한정적이라는 느낌이 들었는데 바뀐 건 범용적으로 느껴지네요 굿! 👍

bimppap commented on src/main/java/cart/dao/CartDao.java:
```diff
@@ -0,0 +1,68 @@
+@Repository
+public class CartDao {
+
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
```
잘 적용해주셨네요! 👍👍

bimppap commented on src/main/java/cart/dao/CartDao.java:
```diff
@@ -0,0 +1,68 @@
+
+    public List<CartData> findAll(final Long userId) {
+        final String sql = "SELECT carts.id, items.name, items.image_url, items.price " +
+                "FROM items JOIN carts " +
```
쿼리에서 조인을 사용할땐 어떤 조인인지 명시해주는게 좋습니다. 조인의 종류에 따라 결과가 다르게 나올 수도 있어요!

bimppap commented on src/main/java/cart/dao/CartDao.java:
```diff
@@ -0,0 +1,68 @@
+        return namedParameterJdbcTemplate.query(sql, param, cartDataRowMapper);
+    }
+
+    public Optional<Cart> findBy(final Long cartId) {
```
메서드명을 이렇게 지은 이유가 궁금하네요! findById라 해도 되지 않나요?

JJ503 replied:
원래 1단계에서는 id로만 찾아 굳이 findBy 뒤에 무엇을 통해를 붙이지 않아도 당연하다고 생각해 생략하게 되었습니다.
또한, 파라미터에 넣어야 하는 값은 Long인 것이 이미 ID를 의미한다 생각해 아이디인 것을 인지할 수 있을 것이라 생각했습니다.

그런데 2단계를 진행하면서 Id 외의 값으로도 검증을 하게 되었기에 Id를 붙여주는 것이 더 명확할 것 같아 수정해 두었습니다!

bimppap commented on src/main/java/cart/domain/CartData.java:
```diff
@@ -0,0 +1,57 @@
+
+import java.util.Objects;
+
+public class CartData {
```
CartData와 Item의 차이는 무엇인가요? 필드가 똑같네요 ㅇ.ㅇ

JJ503 replied:
필드가 똑같긴 하지만, Item의 경우 item 테이블의 해당 행의 id를 CartData의 경우 Cart의 id를 저장하게 됩니다. 그래서 네이밍만 동일할 뿐 서로 다른 데이터를 저장해야 합니다.
또한, 함께 사용하게 된다면 객체 생성 시 헷갈릴 것이라 판단해 따로 만들게 되었습니다.

그런데 결국 Long 데이터를 저장하는 것은 동일하므로 같이 사용하는 게 좋을까요?
혹은 필드 명을 좀 더 명확하게 해주는 것이 좋을까요?
필드명을 그냥 id로 했던 이유는 이미 객체(클래스명)가 어떤 id인지 나타내고 있다고 생각했기 때문입니다…!

bimppap replied:
다른 도메인의 id를 저장하고 있었군요 🙂 현재 CartData는 CartDao에서만 rowMapper 용으로 쓰이고 있습니다. 더 정확히는 findAll에서만 쓰이고 있고요. 그런 점에서 CartData는 도메인보다는 일종의 dto 역할을 하고 있다고 봅니다. 그래서 CartData의 id는 cartId라고 명시를 하는 게 좋을 것 같아요.

CartData를 써도 되지만, 저라면 CartData를 만들지 않고 서비스 레이어에서 기존의 데이터(Cart, Item, User 등)를 가공해 response를 만들 것 같습니다. 정답이 있는 건 아니니 어떻게 사용하면 좋을지 한번 고민해보세요!

bimppap commented on src/main/java/cart/domain/User.java:
```diff
@@ -0,0 +1,88 @@
+
+import java.util.Objects;
+
+public class User {
+
+    private final Long id;
+    private final Email email;
+    private final Password password;
+
```
Spring MVC를 적용하면서 객체지향이 조금 무너지는 경우가 많은데 원시값 포장도 깔끔하게 해주셨네요! 👍👍

bimppap commented on src/main/java/cart/domain/User.java:
```diff
@@ -0,0 +1,88 @@
+        this.password = builder.password;
+    }
+
+    private void validate(Builder builder) {
+        if (builder.email == null) {
+            throw new IllegalArgumentException("이메일은 빈 값일 수 없습니다.");
+        }
+        if (builder.password == null) {
+            throw new IllegalArgumentException("비밀번호는 빈 값일 수 없습니다.");
+        }
+    }
```
검증도 꼼꼼하게 👍👍

bimppap commented on src/main/java/cart/service/ItemService.java:
```diff
@@ -1,40 +1,43 @@
 import java.util.stream.Collectors;
 
 @Service
+@Transactional
```
Transactional 어노테이션을 붙이면 어떤 장점이 있을까요? readOnly=true를 붙이면 또 어떤 차이가 있을까요ㅎㅎ

JJ503 replied:
Service에 Transactional 어노테이션 붙이게 되면 RuntmeException 혹은 오류가 발생한 경우 롤백을 다른 예외가 없을 경우 커밋을 수행해 준다고 알고 있습니다.

readOnly=true에 대한 부분은 공식 문서를 한번 살펴보았습니다.
공식 문서에는 아래와 같이 작성되어 있었습니다.

> This just serves as a hint for the actual transaction subsystem; it will not necessarily cause failure of write access attempts. A transaction manager which cannot interpret the read-only hint will not throw an exception when asked for a read-only transaction but rather silently ignore the hint.
> 

즉, 실제로 readOnly여서 조회 외에는 불가능하게 만드는 것은 아니만, 해당 메서드의 동작에 대한 힌트를 주는 것입니다.
그래서 클래스에 `@Transactional (readOnly=true)`를 붙여주어도 모든 메서드는 수행이 되긴 합니다.

하지만, 읽기에 최적화하므로 변경 여부에 대해 신경을 쓰지 않아 오류가 발생해도 롤백되지 않고 넘어가게 될 수도 있습니다.
대신 읽기에 있어선 트랜잭션 범위 최소화와 같인 불필요한 작업을 줄이기에 보다 성능이 향상된다는 장점이 있습니다.
또한, readOnly=true를 붙임으로써 해당 메서드의 역할의 힌트를 주어 좀 더 직관적으로 알 수 있다는 장점이 있습니다.

그런데 조사하다 보니 조회 외에는 아무 작업도 하지 수행하지 못한다는 이야기가 계속해 나오는데 제가 잘못 이해한 것인지 궁금합니다..!

bimppap replied:
DBSM별로 트랜잭션 어노테이션이 반영되는 방식이 다를 수 있기 때문에 똑같은 어노테이션이어도 다른 결과가 나올 수 있습니다 🙂
관련해서 참고하기 [좋은 글](https://lob-dev.tistory.com/entry/DBMS-%EB%B3%84-Transaction-Read-Only%EC%97%90-%EB%8C%80%ED%95%9C-%EB%8F%99%EC%9E%91-%EB%B0%A9%EC%8B%9D-1)이 있으니 한번 읽어보면 좋을 것 같네요!

bimppap commented on src/main/resources/templates/cart.html:
```diff
@@ -48,9 +49,9 @@ <h1>장바구니</h1>
-                            <img src="" alt=""/> <!-- 상품 이미지 -->
-                            <div class="cart-item-name"></div> <!-- 상품 이름 -->
-                            <div class="cart-item-price"></div> <!-- 상품 가격 -->
+                            <img src="${cartItem.imageUrl}" alt=""/> <!-- 상품 이미지 -->
```
이것만 지우면 완벽한 주석 킬러가 되겠네요 😎

bimppap commented on src/main/java/cart/auth/interceptor/LoginInterceptor.java:
```diff
@@ -0,0 +1,18 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class LoginInterceptor extends HandlerInterceptorAdapter {
```
현재 ArgumentResolver가 LoginInterceptor의 기능도 하고 있다고 보는데 어떻게 생각하시나요? 따로 둔 이유가 있을까요?

JJ503 replied:
각 기능에 대한 제 생각에 대해 간단하게 말씀드리면 우선 `Interceptor`는 특정 url을 요청했을 때 가로채 인증을 수행합니다.
그래서 `ArgumetResolver`를 사용하지 않더라도 해당 페이지를 요청했을 때 header에 있는 Authorization으로 인증되어야 합니다.

반면, `ArgumentResolver`는 요청에 대해서 헤더의 Authorization 값과 같은 정보를 `AuthInfoDto`로 바인딩하는 것 기능을 수행합니다.
그래서 `ArgumentResolver`의 경우는 바인딩을 위해 header에 특정 데이터가 존재하는지만 확인하고 인증은 진행하지 않았습니다.

그런데 제 경우 `Interceptor`에서 실제 계정이 존재하는지에 대한 과정을 수행하지 않았기 때문에 두 기능이 비슷하다고 보였을 수 있겠네요!
`Interceptor`에 계정이 존재하는지에 대한 기능을 추가해 두었습니다! ([로직 바로 가기](https://github.com/woowacourse/jwp-shopping-cart/pull/352/files#r1195149757))

제가 춘식의 질문을 제대로 이해한 것이 맞는지 궁금합니다.

<br>

이와 관련하여 추가적으로 여쭤보고 싶은 부분이 있습니다.
현재 `LoginInterceptor`에서는 해당 계정이 실제 있는지 확인해 줍니다.
그리고 `CartService`에서 장바구니 추가 및 제거, 조회를 수행할 때 다시 해당 계정이 존재하는지 확인하는 작업을 수행해 줍니다. ([로직 바로 가기](https://github.com/woowacourse/jwp-shopping-cart/pull/352/files#r1195153078))
이때, 2차 검증을 수행해 주는 것에 대한 춘식의 의견이 궁금합니다!

현재 일단은 2차 검증을 수행하는 과정으로 두었습니다.
왜냐하면 웹으로 요청을 보내는 경우 `Interceptor`를 통해 인증을 수행할 것이지만, Service를 웹 요청이 아닌 다른 방식으로 사용하게 된다면 검증이 꼭 이루어져야 하기 때문입니다.

그런데 이러한 일이 일어날 수 있음을 가정하는 것이 맞을까요?
혹은 웹 서비스이기에 웹 요청에 대해서만 생각하고 `Interceptor`를 믿고 이후에는 굳이 검증을 하지 않아도 괜찮은 것일까요?

bimppap replied:
ArgumentResolver는 어떠한 요청이 컨트롤러에 들어왔을 때, 요청에 들어온 값으로부터 원하는 객체를 만들어내는 일을 간접적으로 도와주는 기능입니다. 꼭 헤더에서만 가져오는 건 아니에요!

Interceptor 가 인증을 위해 사용되는 것도 맞습니다. 다만 저는 핸들링하는 도메인이 같다면 ArgumentResolver 에서 객체를 만드는 과정에서 검증을 거치고 인터셉터는 생략을 해도 된다고 봐요. ArgumentResolver 가 잘못된 값을 줘서는 안 된다고 생각해서요 🤔 하지만 확장성을 고려하면 역할에 맞게 분리하는 게 좋긴 합니다ㅎㅎ

2차 검증에 대해서는 **없는 것보단 있는 게 낫**다라는 입장입니다! 다른 방식이 어떤 의미인지는 모르겠지만 이상적으로 검증 로직은 꼼꼼할 수록 좋습니다. 하지만 **테스트 코드로 검증 로직 없이도 잘 돌아가는 걸 확신할 수 있다면 제외해도 좋습니다.** 신뢰는 꼼꼼한 검증 로직으로도 얻을 수 있지만 꼼꼼한 테스트로도 얻을 수 있으니까요 👍

bimppap commented on src/test/java/cart/controller/CartControllerTest.java:
```diff
@@ -0,0 +1,146 @@
+import static org.mockito.Mockito.*;
+
+@WebMvcTest(CartController.class)
+class CartControllerTest {
```
<img width="284" alt="image" src="https://user-images.githubusercontent.com/48675973/236921636-3c2cd072-810f-4738-ad95-330ec4f24013.png">

125개... 멋집니다 👍

bimppap reviewed (changes requested):
안녕하세요 제이미, 춘식입니다!
2단계로 오면서 코드가 많이 발전했네요! 꼼꼼히 짠 코드에 감탄하고 갑니다ㅎㅎ 👍
질문에 대한 답과 코멘트 남겼으니 확인 부탁드립니다. 1단계에 추가적으로 남긴 질문에도 답을 달았으니 참고해주세요 🙂
궁금하거나 이해가 안 되는 점은 편하게 DM 주세요.

JJ503 committed:
feat: AuthorizationException 예외 처리 기능 추가

JJ503 committed:
refactor: AuthorizationException 예외 메시지 추가

JJ503 commented on src/main/java/cart/dao/CartDao.java:
```diff
@@ -0,0 +1,68 @@
+public class CartDao {
+
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    private final RowMapper<Cart> cartRowMapper = (resultSet, rowNumber) -> new Cart.Builder()
+            .id(resultSet.getLong("id"))
+            .userId(resultSet.getLong("user_id"))
+            .itemId(resultSet.getLong("item_id"))
+            .build();
+    private final RowMapper<CartData> cartDataRowMapper = (resultSet, rowNumber) -> new CartData(
+            resultSet.getLong("id"),
+            new Name(resultSet.getString("name")),
+            new ImageUrl(resultSet.getString("image_url")),
+            new Price(resultSet.getInt("price"))
+    );
```
dao에서 서로 다른 도메인을 반환하는 경우
[리뷰 답변 참고](https://github.com/woowacourse/jwp-shopping-cart/pull/352#discussion_r1194467341)

JJ503 committed:
refactor: 불필요한 주석 제거

JJ503 committed:
refactor: BasicAuthenticationPrincipalArgumentResolver 빈으로 등록해 사용

JJ503 committed:
refactor: AuthInfoDto 클래스명 수정

AuthInfo -> AuthInfoDto

JJ503 committed:
refactor: LoginInterceptor 계정이 존재하는지 검증 기능 추가

- Userdao 이메일과 비밀번호로 사용자를 찾는 기능 추가
- UserService 특정 이메일 비밀번호를 갖는 사용자가 존재하는지 확인하는 기능 추가
- BasicAuthorizationExtractor 예외 메시지 추가
- UserDao 단위 테스트 추가
- UserService 단위 테스트 추가

JJ503 committed:
refactor: findBy() 메서드명을 findById()로 명확하게 수정

JJ503 committed:
fix: 테스트 오류 발생 해결

JJ503 committed:
refactor: 쿼리문 join 종류 명시

JJ503 committed:
refactor: RestAssured 테스트 수행 시 로그를 남기도록 수정

JJ503 committed:
refactor: 완료한 todo 주석 제거

JJ503 committed:
feat: 인증 값의 유효성 검사 기능 추가

JJ503 commented on src/main/java/cart/config/WebMvcConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+@Configuration
+public class WebMvcConfiguration implements WebMvcConfigurer {
+
+    private final LoginInterceptor loginInterceptor;
+    private final BasicAuthenticationPrincipalArgumentResolver basicAuthenticationPrincipalArgumentResolver;
+
+    public WebMvcConfiguration(LoginInterceptor loginInterceptor,
+                               BasicAuthenticationPrincipalArgumentResolver basicAuthenticationPrincipalArgumentResolver) {
+        this.loginInterceptor = loginInterceptor;
+        this.basicAuthenticationPrincipalArgumentResolver = basicAuthenticationPrincipalArgumentResolver;
+    }
```
빈 등록을 통한 의존성 주입
[리뷰 답변 참고](https://github.com/woowacourse/jwp-shopping-cart/pull/352#discussion_r1195136967)

JJ503 commented on src/main/java/cart/auth/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,54 @@
+        return null;
+    }
+
+    private void isValidBase64(String authHeaderValue) {
+        String exceptionMessage = "잘못된 인증 정보입니다.";
+        if (authHeaderValue.split(" ").length > 1) {
+            throw new AuthorizationException(exceptionMessage);
+        }
+        if (!authHeaderValue.matches(BASE64_REGEX)) {
+            throw new AuthorizationException(exceptionMessage);
+        }
+        if (authHeaderValue.length() % 3 != 0) {
+            throw new AuthorizationException(exceptionMessage);
+        }
+    }
```
Base64 인코딩 값 유효성 검사
[리뷰 답변 참고](https://github.com/woowacourse/jwp-shopping-cart/pull/352#discussion_r1195141035)

JJ503 commented on src/main/java/cart/auth/interceptor/LoginInterceptor.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
+        AuthInfoDto authInfoDto = basicAuthorizationExtractor.extract(request);
+        userService.isExistUser(authInfoDto);
```
`Interceptor`
- 계정이 존재하는지에 대한 검증 추가

[리뷰 답변 참고](https://github.com/woowacourse/jwp-shopping-cart/pull/352#discussion_r1187838485)

JJ503 commented on src/main/java/cart/service/CartService.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    public Long saveCart(final AuthInfoDto authInfoDto, final Long itemId) {
+        Long userId = findUserIdByEmail(authInfoDto);
+        validateExistItem(itemId);
+        Cart cart = new Cart.Builder()
+                .userId(userId)
+                .itemId(itemId)
+                .build();
+        return cartDao.save(cart);
```
`CartService`
- 사용자, 아이템 존재 여부에 대한 2차 검증

[리뷰 답변 참고](https://github.com/woowacourse/jwp-shopping-cart/pull/352#discussion_r1195145484)

JJ503 commented:
안녕하세요 춘식!
2단계 리뷰에 대한 피드백 적용 후 리뷰 재요청드립니다.

생각보다 리뷰 요청이 더 늦어져 죄송할 따름이네요... 😥
편하실 때 리뷰 해주시면 감사하겠습니다..!

또한, 춘식이 남겨준 질문들과 추가적인 의문점들에 대해선 코멘트로 작성해 두었습니다.
확인해 주시면 감사하겠습니다.

이번 리뷰도 잘 부탁드립니다 😊

bimppap reviewed (approved):
안녕하세요 제이미, 춘식입니다!
제가 일이 많아서 답이 늦었네요ㅠㅠ리팩토링 깔끔하게 잘 해주셨어요!
피드백에 대해 많이 찾아보고 고민한 게 느껴져서 저도 공부가 된 것 같네요ㅎㅎ
장바구니 미션은 여기서 마무리하도록 하겠습니다. 고생하셨습니다!
질문에 대한 답에 코멘트들 짧게 달아놨으니 확인 부탁드립니다.
남은 미션들도 파이팅입니다! 👍
